### PR TITLE
Run each console command in its own child process

### PR DESCRIPTION
**Pull request by fainashalts**
This PR runs commands from inside the `truffle develop` and `truffle console` command processes. While this PR changes the way `truffle develop` and `truffle console` work under the hood, the end user CLI experience should not be different. 

These changes became necessary when we found that Node v12 does not allow `UncaughtException` listeners in a REPL within another NodeJS program, and that one of our dependencies uses such a listener. By running a separate REPL in a child process we are able to bypass this problem. The issue that brought this to our attention is here: https://github.com/trufflesuite/truffle/issues/2463

In the process of this work, it became apparent that the ReplManager class we had been using to manage multiple REPLs was no longer necessary since each REPL will now be in its own process. This PR thus also removes the `repl.js` file and directly accesses the node REPL as needed in `console.js` and `interpreter.js`. 

To test this PR, run either `truffle develop` or `truffle console` and then use these to test out any other Truffle commands that make sense: `migrate --reset`, `compile`, `test`, `debug`, etc. Everything should work as it did previously, with the exception that the `--network` flag will throw an error when used with `truffle develop` -- the only network that should be used by `truffle develop` is the `develop` network.

### CODE REVIEW

File: packages/core/index.js [hunk 1]
```diff
@@ -4,7 +4,6 @@ var pkg = require("./package.json");
 module.exports = {
   build: require("./lib/build"),
   create: require("./lib/commands/create/helpers"),
-  console: require("./lib/repl"),
```
Comment (fainashalts) [inline packages/core/index.js:7]: @gnidan I removed this because it wasn't being used anywhere, but now I see it is in the truffle scenario tests, which are failing in GHA. Removing this test solves it, but I wanted to check in first in case I'm missing something: https://github.com/trufflesuite/truffle/blob/develop/packages/truffle/test/scenarios/library/api.js#L26
Ok to remove?

File: packages/core/lib/console.js [hunk 2]
```diff
@@ -143,16 +126,53 @@ class Console extends EventEmitter {
...
+      );
+
+      try {
+        this.provision();
+      } catch (e) {
+        console.log(e);
+      }
+    } catch (err) {
+      callback(err);
+    }
```
Comment (cds-amal) [inline packages/core/lib/console.js:161]: Is the intention to eat exceptions thrown by `this.provision()` and not invoke the callback on 160?

File: packages/core/index.js (hunk 1 above)
Comment (eggplantzzz) [reply packages/core/index.js:7]: No, I think this is being exposed so that users can pull it out and use it separately.

File: packages/core/lib/debug/interpreter.js [hunk 3]
```diff
@@ -304,10 +293,7 @@ class DebugInterpreter {
     }
 
     //split arguments for commands that want that; split on runs of spaces
-    splitArgs = cmd
-      .trim()
-      .split(/ +/)
-      .slice(1);
+    splitArgs = cmd.trim().split(/ +/).slice(1);
```
Comment (cds-amal) [inline packages/core/lib/debug/interpreter.js:296]: Note this does not account for a quoted argument with spaces. I know this isn't part of the work done in the PR but wanted to surface it

File: packages/core/lib/console-child.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+const TruffleError = require("@truffle/error");
+const Config = require("@truffle/config");
+const Web3 = require("web3");
+
+const inputStrings = process.argv[2];
+//deserialize options
+const options = JSON.parse(process.argv[3]);
+
+//deserialize networks entries
+for (var key in options.networks) {
```
Comment (eggplantzzz) [inline packages/core/lib/console-child.js:11]: Can we use `const key`?

File: packages/core/lib/console-child.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  options.networks[key] = JSON.parse(options.networks[key]);
+}
+
+//detect config so we can get the provider and resolver without having to serialize
+//and deserialize them
+const configDetected = Config.detect(options);
```
Comment (eggplantzzz) [inline packages/core/lib/console-child.js:17]: What about naming this `detectedConfig`? `configDetected` sort of sounds like a predicate.

File: packages/core/lib/console.js [hunk 6]
```diff
@@ -143,16 +126,53 @@ class Console extends EventEmitter {
...
+
+    const spawnOptions = { stdio: ["inherit", "inherit", "inherit"] };
 
-    this.repl.setContextVars(contextVars);
+    try {
+      //serialize config networks entries
+      const configNetworks = {};
+      for (const key in options.networks) {
+        configNetworks[key] = JSON.stringify(options.networks[key]);
+      }
```
Comment (cds-amal) [inline packages/core/lib/console.js:146]: How many network properties are used in the child side? If it's 2-5 can this be changed to explicitly name them as arguments to spawnSync? A benefit of this approach would be to document what props are being used.
```
const networkProps = [ '--networkProp1', value1, '--networkProp2', 'value2'...];
SpawnSync("node", [...otherArgs, ...networkProps])
```

File: packages/core/lib/console.js [hunk 7]
```diff
@@ -143,16 +126,53 @@ class Console extends EventEmitter {
     abstractions.forEach(abstraction => {
       contextVars[abstraction.contract_name] = abstraction;
     });
+  }
+
+  runSpawn(inputStrings, options, callback) {
+    let args;
```
Comment (cds-amal) [inline packages/core/lib/console.js:132]: `args` is confusing in this context because It implies some none zero number of arguments when, in this case, it is the module/entryPoint to spawn

File: packages/core/index.js (hunk 1 above)
Comment (fainashalts) [reply packages/core/index.js:7]: So I wound up taking out the test, after discussing with gnidan. The repl manager doesn't exist anymore so that export isn't really meaningful anymore. Let me know if you want to discuss this more!

File: packages/core/lib/console.js [hunk 8]
```diff
@@ -54,35 +56,16 @@ class Console extends EventEmitter {
...
+      this.interfaceAdapter.getAccounts().then(() => {
         const abstractions = this.provision();
 
-        this.repl.start({
+        this.repl = repl.start({
           prompt: "truffle(" + this.options.network + ")> ",
-          context: {
-            web3: this.web3,
-            interfaceAdapter: this.interfaceAdapter,
-            accounts: fetchedAccounts
```
[the remark this answers is not in the document]
Comment (fainashalts) [reply packages/core/lib/console.js:82]: Totally right. Fixed!

File: packages/core/lib/console.js (hunk 2 above)
Comment (fainashalts) [reply packages/core/lib/console.js:165]: Unless I'm misunderstanding, the callback on lime 160 is for any exceptions thrown by spawning the repl. There's a try/catch for `this.provision()` and a try/catch for the spawning. Does that make sense or am I not following?

File: packages/core/lib/debug/interpreter.js (hunk 3 above)
Comment (fainashalts) [reply packages/core/lib/debug/interpreter.js:296]: @cds-amal can you make a separate issue for this?

File: packages/core/lib/console.js (hunk 7 above)
Comment (fainashalts) [reply packages/core/lib/console.js:136]: Agree. Renamed to `childPath` to be more clear.

File: packages/core/lib/console-child.js (hunk 4 above)
Comment (fainashalts) [reply packages/core/lib/console-child.js]: I removed this part of the code, so no need! 🎉

File: packages/core/lib/console-child.js (hunk 5 above)
Comment (fainashalts) [reply packages/core/lib/console-child.js]: Agree. Changed.

**Comment by gnidan** [comment]
Hey @fainashalts, check out the bundled version of Truffle - there's a deprecation notice in every child process:

```
truffle(develop)> test
(node:96195) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Using network 'develop'.

```

File: packages/core/lib/console.js [hunk 9]
```diff
@@ -143,16 +130,53 @@ class Console extends EventEmitter {
...
+
+    const spawnOptions = { stdio: ["inherit", "inherit", "inherit"] };
 
-    this.repl.setContextVars(contextVars);
+    try {
+      //serialize config networks entries
+      const configNetworks = {};
+      for (const key in options.networks) {
+        configNetworks[key] = JSON.stringify(options.networks[key]);
+      }
```
Comment (gnidan) [inline packages/core/lib/console.js:150]: `configNetworks` appears to be unused

File: packages/core/lib/debug/interpreter.js (hunk 3 above)
Comment (gnidan) [reply packages/core/lib/debug/interpreter.js:296]: note that this is the debug repl, I'm not sure what behavior you're describing @cds-amal, but the use case might be different here.

Comment (cds-amal) [reply packages/core/lib/debug/interpreter.js:296]: Oh. If there's an expectation for no quoted arguments then all is good. The parsing behavior I tried to explain in the following block.  

```
> cmd = "command arg1 \"arg two\" arg3"
'command arg1 "arg two" arg3'
> cmd.trim().split(/ +/).slice(1)
[ 'arg1', '"arg', 'two"', 'arg3' ]

```

File: packages/core/lib/console.js (hunk 2 above)
Comment (cds-amal) [reply packages/core/lib/console.js:165]: The intention is a bit hard to follow because the two catch expressions are next to each other.  As it exists, if the provision after spawn fails then the code explicitly eats the error and continues execution to the next REPL prompt. Is an exception thrown by `this.provision()` benign as the code suggests?

File: packages/core/lib/debug/interpreter.js (hunk 3 above)
Comment (gnidan) [reply packages/core/lib/debug/interpreter.js:296]: Yep, the debugger doesn't really provide that level of sophistication! (Maybe it should? But I think we can ignore for now)

File: packages/core/lib/console-child.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
+const Command = require("../lib/command");
+const TruffleError = require("@truffle/error");
+const Config = require("@truffle/config");
+const Web3 = require("web3");
+
+const inputStrings = process.argv[2];
+const network = process.argv[3];
```
Comment (gnidan) [inline packages/core/lib/console-child.js:7]: Can we merge these two? i.e., if the user runs `debug --network mainnet` from inside `truffle console --network development`, then I would expect the `debug` command to connect to mainnet. This implementation suggests that behavior might be undefined (or favor the network specified by `process.argv[3]`)

Seems like we shouldn't pass `network` separately as a positional argument, and instead pass it as a named `--network` option. Two options come to mind for how to do that:

1. Add logic to the parent process to merge `--network=<network-name>` into `inputStrings`, possibly by decoding `inputStrings` into a mapping + then re-encoding with `{ network }` merged in.

2. Use a ` -- ` separator between parent-process-provided options and user-provided options, e.g.:
  ```
  truffle(develop)> debug <tx-hash> --fetch-external
     ... would run command:
     $ console-child.js --network=develop -- <tx-hash> --fetch-external
  ```

Comment (fainashalts) [reply packages/core/lib/console-child.js:7]: Hi @gnidan!  https://github.com/trufflesuite/truffle/pull/3296 addresses this in case you do want to get rid of the positional argument. Wanted to save you a little time!

Comment (gnidan) [reply packages/core/lib/console-child.js:7]: Thanks so much @fainashalts, thanks for knocking that out!

That PR looks great! But why is it failing ☹️ ? (Re-running now)

File: packages/core/lib/console.js (hunk 2 above)
Comment (cds-amal) [reply packages/core/lib/console.js:161]: `this.provision();` can throw an Error (line 110) if something is awry with an artifact file. I think we should be invoke the callback in the outer catch block.
